### ISSUE TITLE
test_haiku fails with NullMask (mask rules out all tokens) — not backend-specific

### ISSUE THREAD

**Issue by postylem** [genlm/llamppl#47]
## What
`tests/test_examples.py::test_haiku[vllm]` fails under the **vLLM** backend:

```
llamppl.distributions.lmcontext.NullMask: Unable to compute log probability of mask that rules out all tokens.
```

The haiku example reaches a step where the constraint masks out *every* token, so the log-probability of the mask can't be computed.

## Context
- Surfaced once GPU CI was restored (the partner GPU runner had been auto-cancelling jobs at GitHub's 24h queue cap) and vLLM was correctly installed in the `build` job (#46).
- The other `[vllm]` tests pass: `test_examples.py::test_hard_constraints[vllm]` and `test_lmcontext.py::test_init[vllm]`. So vLLM itself works — this is specific to the haiku example under vLLM.
- Backend: `gpt2` via `genlm-backend[vllm]` (vLLM 0.10.x), `n_particles=20`.

## Current state
Scoped, non-strict **xfail** on the `[vllm]` param only (added in #46) so it doesn't block CI:
in `tests/test_examples.py::test_haiku`, `request.applymarker(pytest.mark.xfail(...))` when `backend == "vllm"`.

## Possible directions
- Could be sampling-dependent (re-run to check determinism) or a genuine edge case where the haiku syllable/structure constraint admits no continuation given gpt2's vLLM logits/tokenization.
- If real, the haiku example and/or `NullMask` handling in `llamppl/distributions/lmcontext.py` may need to handle the all-tokens-masked state gracefully.

**When fixed:** remove the xfail in `tests/test_examples.py::test_haiku`.

**Comment by postylem** [comment]
Update from local investigation. This is **not vLLM-specific**

- Reproduced on the **hf** backend locally (~1 `NullMask` in ~30 runs of the haiku example: gpt2, n_particles=20). Also seen on vLLM in CI in 2 runs.
- The **mock** backend never failed (0/30).

Asked to investigate, Claude wrote:

**Root cause:** `sample_word` (`llamppl/chunks.py`) forces a word boundary via `observe(context.mask_dist(STARTS_NEW_WORD | CONTINUES_CURRENT_WORD), ...)`. `Mask.log_prob` (`llamppl/distributions/lmcontext.py`) raises `NullMask` when the conditioned mask has zero overlap with the current `model_mask` — i.e. conditioning on a zero-probability event. Instead of giving that particle weight 0 (the correct SMC behaviour), it raises and crashes the whole inference.

**Secondary:** a `NullMask` raised mid-SMC leaves a reused `LLM`'s async state corrupted (subsequent runs on the same object fail with `asyncio.InvalidStateError`); a fresh `LLM` per run doesn't show this.

**Note:** the failure is stochastic (the test is unseeded). All sampling uses the global numpy RNG, so `np.random.seed(s)` before a run makes it deterministic — useful for a deterministic repro/regression test. `tests/test_examples.py::test_haiku[vllm]` is currently xfailed (non-strict) to keep CI green.

**Comment by postylem** [comment]
Opened #50. The fix mirrors `condition(False)`: `LMTokenMask.log_prob` returns `-inf` for an impossible mask (instead of raising `NullMask`), and `Model.observe` finishes the particle on a `-inf` observation. So the impossible particle is given weight 0 and dropped at the next resample, rather than aborting the run. The handling lives in `observe` and `log_prob`, so it applies to all inference drivers. The unused `NullMask` exception is removed, and the `test_haiku[vllm]` xfail is removed.
